---
title:  "p28443 PacBio bam and QC"
author: "Natalia Zajac"
output:
  html_document: 
    highlight: pygments
    theme: sand
    code_folding: hide
    toc: yes
    toc_depth: 4
editor_options: 
  chunk_output_type: inline
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup, include=FALSE}
library(SummarizedExperiment)
library(plotly)
library(scater)
library(ezRun)
library(DT)
library(htmltools)
library(Matrix)
library(Seurat)
library(pheatmap)
library(DropletUtils)
library(cowplot)
library(tidyverse)
library(ggvenn)
library(patchwork)
library(UpSetR)
library(clusterProfiler)
library(org.Hs.eg.db)
library(BiocGenerics)
library(cowplot)
library(gridExtra)
library(ggpubr)

source("functions.R")
sample_data = ezRead.table("/home/zajac/Genomics/p28443_singlecell_pacbiovsillumina/SIB-poster/2024-05-22-sampleInfo.tsv")
##Load all pacbio dirs

geneAnno <- ezRead.table("/srv/GT/reference/Homo_sapiens/GENCODE/GRCh38.p13/Annotation/Release_39-2021-12-09/Genes/SMRTLink11.1_annotation/genes_annotation_byGene.txt")
#Read in mtcontent from Hubert


sample_data$pbqs = file.path("/srv/GT/analysis/zajacn/p28443/subset_bams_072024", list.files("subset_bams_072024/", pattern = "-pbGa.qsd"))[c(5,1,2,3,4)]
sample_data$illqs = file.path("/srv/GT/analysis/zajacn/p28443/subset_bams_072024", list.files("subset_bams_072024/", pattern = "-illGa.qsd"))[c(5,1,2,3,4)]
sample_data$illqsUn = file.path("/srv/GT/analysis/zajacn/p28443/subset_bams_072024", list.files("subset_bams_072024/", pattern = "-illUnmapped.qsd"))[c(5,1,2,3,4)]
sample_data$umicounts = file.path("/srv/GT/analysis/zajacn/p28443/UMI_summary", list.files("UMI_summary/", pattern = "-cellUMIStats.qsd"))[c(5,1,2,3,4)]
sample_data$tsocounts = file.path("/srv/GT/analysis/zajacn/p28443/UMI_summary", list.files("UMI_summary/", pattern = "tso_pa.qsd"))[c(5,1,2,3,4)]
sample_data$mtcontent = file.path("/srv/GT/analysis/zajacn/p28443/UMI_summary", list.files("UMI_summary/", pattern = "-cellMITOStats.qsd"))[c(5,1,2,3,4)]

mtcontent = lapply(sample_data$mtcontent, function(qsFile){
  qsFile %>% qs::qread() 
})
names(mtcontent) <- sample_data$Name
pbGaList <- lapply(sample_data$pbqs, function(qsFile){
  qsFile %>% qs::qread() %>% granges(use.mcols=TRUE)
})
names(pbGaList) <- sample_data$Name
illGaList <- lapply(sample_data$illqs, function(qsFile){
  qsFile %>% qs::qread() %>% granges(use.mcols=TRUE)
})
names(illGaList) <- sample_data$Name
illUnmappedList <- lapply(sample_data$illqsUn, qs::qread)
names(illUnmappedList) <- sample_data$Name

nCommonCellsUsed <- lapply(sample_data$Name, function(sm){
  intersect(mcols(pbGaList[[sm]])$CB, mcols(illGaList[[sm]])$CB) %>% unique() %>% length()
})
names(nCommonCellsUsed) <- sample_data$Name

UMIlist <- lapply(sample_data$umicounts, function(qsFile){
  qsFile %>% qs::qread()
})
names(UMIlist) <- sample_data$Name

# Filter duplicate reads in Illumina
illGaDF = lapply(illGaList, function(x){
  df = x[x$xf == "25",] #choose the mapped and counted
  dfrest = x[!x$xf %in% c("25", "0"),] #exclude mapped and counted and unmapped, deduplicate keeping one at random and not those that are already present at df
  dfrest = dfrest[!duplicated(dfrest$tagId),]
  dfrest = dfrest[!dfrest$tagId %in% df$tagId]
  dfnull = x[x$xf == "0",] #choose unmapped
  dfnull = dfnull[!dfnull$tagId %in% c(dfrest$tagId, df$tagId),] #kep only those that are not present already
  final = makeGRangesFromDataFrame(rbind(data.frame(df), data.frame(dfrest), data.frame(dfnull)), keep.extra.columns = T)
  return(final)
})


# Filter duplicate reads in PacBio
pbGaDF = lapply(pbGaList, function(x){ 
  x = x[!duplicated(x$tagId)]
})

#Load TSO data
tso = lapply(sample_data$tsocounts, function(qsFile){
  qsFile %>% qs::qread()
})
names(tso) = sample_data$Name
```


Number of reads

```{r, fig.width=6, fig.height=5, echo=FALSE, message=FALSE, warning=FALSE}
x = melt(sample_data[,c(1,8,10),] %>% rename(Illumina = nIlluminaReads, PacBio = nPbSegmentReads)) 
x$Name = factor(x$Name, levels = sample_data$Name)
x$value = x$value / 1000000
p1 = ggplot(x, aes(Name, value, fill = variable)) + 
  geom_bar(position = "dodge", stat = "identity") + 
  scale_fill_manual(values = myColors) + 
  labs(x = "Sample", y = "Number of reads [mio]", fill = "") + theme(legend.position = "top", text = element_text(size = 15))
p1
```

UMI summary

```{r, fig.width=10, fig.height=8,echo=FALSE, message=FALSE, warning=FALSE}
UMIs = bind_rows(UMIlist, .id = "Sample")
UMIs$Sample = factor(UMIs$Sample, levels = sample_data$Name)
UMIs$UMI_count_Ill = UMIs$UMI_count_Ill/1000
UMIs$UMI_count_Pb = UMIs$UMI_count_Pb/1000

p1 = ggplot(UMIs[UMIs$Commonness == "Common",], aes(UMI_count_Pb, UMI_count_Ill, color = Sample)) + geom_point() + xlim(0,500) + ylim(0,500) + geom_abline(linetype = "dotted") + scale_color_manual(values = sampleColors) + labs(x = "UMI count per cell PacBio[x10^3]", y = "UMI count per cell Illumina [x10^3]") + stat_cor(method = "pearson", label.x.npc = "middle", label.y.npc = "top", size = 8) + theme(text = element_text(size = 20)) 
p1
```

UMIs in Illumina cells

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, warning=FALSE}
library(ggbeeswarm)
p4 = ggplot(UMIs[UMIs$Commonness %in% c("Unique to ILL", "Common"),] %>% mutate(Commonness = if_else(Commonness == "Unique to ILL", "Illumina specific", Commonness)), aes(Sample, UMI_count_Ill, fill = Commonness )) + geom_violin() + geom_quasirandom( aes(color = Commonness), dodge.width = 1) + scale_fill_manual(values = myColors) + scale_color_manual(values = myColors) + labs(fill = "", color = "", y = "UMI count per cell[x10^3]") + theme(legend.position = "top", text = element_text(size = 8), legend.text = element_text(size = 13))
p4
```

```{r, fig.width=10, fig.height=8, echo=FALSE, message=FALSE, warning=FALSE}
p1 + annotation_custom(ggplotGrob(p4), xmin = 200, xmax = 500, 
                       ymin = 0, ymax = 185)
```

Total UMI count per sample

```{r, fig.width=6, fig.height=5, echo=FALSE, message=FALSE, warning=FALSE}
UMIs$TotalPB = UMIs$TotalPB/1000000
UMIs$TotalIll = UMIs$TotalIll/1000000

p3 = UMIs[,c(1,3,5)] %>% rownames_to_column() %>% dplyr::select("Sample", "TotalPB", "TotalIll") %>% unique() %>% filter(TotalPB > 0) %>% melt() %>% mutate(variable = if_else(variable == "TotalPB", "PacBio", "Illumina")) %>% ggplot(aes(Sample, value, fill = variable)) + geom_bar(position = "dodge", stat = "identity") + scale_fill_manual(values = myColors) + labs(fill = "", y = "Total Number of UMIs [mio]") + theme(legend.position = "top", text = element_text(size = 15)) 
p3
```

Number of cells

```{r, fig.width=6, fig.height=5, echo=FALSE, message=FALSE, warning=FALSE}
df = UMIs %>% rownames_to_column("CB") %>% group_by(Sample, Commonness) %>% summarise(Cell_count = n_distinct(CB))
df = df %>% pivot_wider(names_from = Commonness, values_from = Cell_count) %>% mutate("Illumina specific" = Common + `Unique to ILL`, "PacBio specific" = Common + `Unique to PB`) %>% replace(is.na(.), 0) %>% dplyr::select(Sample, Common, `Illumina specific`, `PacBio specific`) 

p2 = ggplot() + 
    geom_bar(data = melt(df[,c(1,3,4)]), aes(Sample, value, fill = variable), position = "dodge", stat = "identity", show.legend = TRUE) + 
    geom_bar(data=melt(df[c(1,2)]), stat="identity", aes(Sample, value, fill = variable), show.legend = TRUE) +
    scale_fill_manual(name = "", values = myColors) + labs(y = "Number of Cells") + theme(legend.position = "top", text = element_text(size = 15))
p2
```

Number of reads per tag ID

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, warning=FALSE}
for (i in sample_data$Name){
  print(data.frame(illGaList[[i]]) %>% group_by(tagId) %>% dplyr::summarise(count = n_distinct(qname)) %>% summarise(mean(count)))
}
for (i in sample_data$Name){
  print(data.frame(pbGaList[[i]]) %>% group_by(tagId) %>% dplyr::summarise(count = n_distinct(qname)) %>% summarise(mean(count)))
}
```

Unmapped and illStatus

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, warning=FALSE}
for (i in sample_data$Name){
  print(data.frame(pbGaDF[[i]]) %>% group_by(illUnmapped, illStatus) %>% dplyr::summarise(mean(gc), mean(qlength), mean(transcriptLength), count = n_distinct(tagId)) %>% ungroup() %>% mutate(prop = (count *100)/ sum(count)))
}
```

What are the categories of the illAbsent reads

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, warning=FALSE}
for (i in sample_data$Name){
  print(data.frame(pbGaDF[[i]][pbGaDF[[i]]$illStatus == "illAbsent",]) %>% group_by(category) %>% dplyr::summarise(count = n()) %>% ungroup() %>% mutate(prop = (count *100)/ sum(count))%>% filter(category %in% c("full-splice_match", "incomplete-splice_match")) %>% dplyr::summarise(sum(prop)))
}
```

What are the categories of the illDetected reads

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, warning=FALSE}
for (i in sample_data$Name){
    print(data.frame(pbGaDF[[i]][pbGaDF[[i]]$illStatus == "illDetected",]) %>% group_by(category) %>% dplyr::summarise(count = n()) %>% ungroup() %>% mutate(prop = (count *100)/ sum(count))%>% filter(category %in% c("full-splice_match", "incomplete-splice_match")) %>% dplyr::summarise(sum(prop)))
}
```


Mitochondrial content

```{r, fig.width=10, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}

p1 = ggplot(bind_rows(mtcontent, .id = "Sample") %>% filter(!is.na(pbPercentMito)), aes(pbPercentMito, illPercentMito, color = factor(Sample, levels = sample_data$Name))) + 
  geom_point() + 
  scale_color_manual(values = sampleColors) + 
  labs(x = "Mitochondrial Content per cell PacBio", y = "Mitochondrial Content per cell Illumina", color = "Sample") + stat_cor(method = "pearson", size = 8) + theme(text = element_text(size = 20))
p1
```

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, warning=FALSE}
mtill = bind_rows(mtcontent, .id = "Sample") %>% mutate(inPB = if_else(is.na(pbPercentMito), "Illumina specific", "Common"))
mtill$Sample = factor(mtill$Sample, levels = unique(sample_data$Name))
p2 = ggplot(mtill, aes(Sample, illPercentMito, fill = inPB)) + geom_violin() + geom_quasirandom(aes(color = inPB), dodge.width = 1) + scale_fill_manual(values = myColors) + scale_color_manual(values = myColors) + labs(y = "Mitochondrial content %", fill = "", color = "") + theme(text = element_text(size =8), legend.position = "top", legend.text = element_text(size = 13))
p2
```


```{r, fig.width=10, fig.height=8, echo=FALSE, message=FALSE, warning=FALSE}
p1 +  annotation_custom(ggplotGrob(p2), xmin = 50, xmax = 100, 
                         ymin = 0, ymax = 35)
```

Shared tags summary

```{r, fig.width=6, fig.height=5, echo=FALSE, message=FALSE, warning=FALSE}
tagIds = NULL
for (i in sample_data$Name){
  x = data.frame(pbGaDF[[i]]) %>% group_by(illStatus) %>% dplyr::summarise(count = n_distinct(tagId)) %>% mutate(Sample = i) %>% rename("Status" = "illStatus")
  y = data.frame(illGaDF[[i]]) %>% group_by(pbStatus) %>% dplyr::summarise(count = n_distinct(tagId)) %>% mutate(Sample = i) %>% rename("Status" = "pbStatus")
  tagIds = rbind(tagIds, rbind(x, y))
}
tagIds = tagIds %>% mutate(Status = if_else(Status %in% c("illDetected", "pbDetected"), "Common", Status)) %>% unique() %>% pivot_wider(names_from = Status, values_from = count) %>% mutate("PacBio" = illAbsent + Common, "Illumina" = pbAbsent + Common) %>% dplyr::select(Sample, Common, PacBio, Illumina)
tagIds$Sample = factor(tagIds$Sample, levels = sample_data$Name)
tagIds[,c(2,3,4),] = tagIds[,c(2,3,4),]/1000000

ggplot() + 
    geom_bar(data = melt(tagIds[c(1,3,4)]), aes(Sample, value, fill = factor(variable, levels = c("Illumina", "PacBio", "Common"))), position = "dodge", stat = "identity", show.legend = TRUE) + 
    geom_bar(data=melt(tagIds[c(1,2)]), stat="identity", aes(Sample, value, fill = variable), show.legend = TRUE) +
    scale_fill_manual(name = "", values = myColors, labels=c("Illumina","PacBio", "Common")) +
    labs(y = "Number of Cell Barcode-UMI tags [mio]") + theme(legend.position = "top", text = element_text(size = 15))

#geom_text(data = melt(shareddf[c(1,2,5)], id.vars = c("Percentage", "Sample")), aes(Sample, value - 400, label = paste0(Percentage, "%")), color="white", size=3.5)
```


```{r, fig.width=25, fig.height=10, echo=FALSE, message=FALSE, warning=FALSE}
multimappers = NULL
for (i in sample_data$Name){
  x = data.frame(pbGaDF[[i]][pbGaDF[[i]]$illStatus == "illDetected"]) %>% 
    group_by(multiMapping, quality) %>% 
    dplyr::summarise(count = n_distinct(tagId)) %>% 
    mutate(quality = if_else(quality == "Pass", "Counted", "NotCounted")) %>% 
    group_by(multiMapping, quality) %>% 
    dplyr::summarise(count = sum(count)) %>% 
    mutate(Sample = i, Method = "PacBio")
  x$Percentage = (x$count* 100)/sum(x$count)
  
  y = data.frame(illGaDF[[i]][illGaDF[[i]]$pbStatus == "pbDetected",]) %>% 
    group_by(NH, umiCount) %>% 
    dplyr::summarise(count = n_distinct(tagId)) %>% 
    mutate(multiMapping = if_else(NH == 1, "Uniquely mapping", "Multi mapping"), quality = umiCount) %>% 
    group_by(multiMapping, quality) %>% 
    dplyr::summarise(count = sum(count)) %>% 
    mutate(Sample = i, Method = "Illumina", Percentage = (count * 100)/sum(count))
  y$Percentage = (y$count* 100)/sum(y$count)
  multimappers = rbind(multimappers, rbind(x,y))
}

discarded = NULL
for (i in sample_data$Name){
  x = data.frame(pbGaDF[[i]][pbGaDF[[i]]$illStatus == "illDetected"])
  x = x[x$multiMapping == "Uniquely mapping",]
  x = x %>% group_by(quality, multiMapping) %>% dplyr::summarise(count = n_distinct(tagId))
  x$quality2 = x$quality
  x = x %>% group_by(quality2) %>% dplyr::summarise(total_count = sum(count), .groups = 'drop')
  x = x %>% mutate(proportion = total_count / sum(total_count)) %>% ungroup()
  x$Sample = i
  x$Method = "PacBio"
  x$quality2 = if_else(x$quality2 ==  "Pass", "Counted Uniquely mapping", x$quality2)
  discarded = rbind(discarded, x)
}

sub_data = NULL
for (i in sample_data$Name){
  x = data.frame(pbGaDF[[i]][pbGaDF[[i]]$illStatus == "illDetected"]) %>% 
    group_by(multiMapping, quality) %>% 
    dplyr::summarise(count = n_distinct(tagId)) %>% 
    mutate(quality2 = case_when(quality == "Pass" ~ "Counted Uniquely mapping", quality == "LowCoverage/Non-Canonical" ~  "LCNC", .default = quality)) %>% 
    mutate(quality = if_else(quality == "Pass", "Counted", "NotCounted")) %>% 
    group_by(multiMapping, quality, quality2) %>% 
    dplyr::summarise(count = sum(count)) %>% 
    mutate(Sample = i, Method = "PacBio")
  x$proportion = (x$count* 100)/sum(x$count)
  sub_data = rbind(sub_data, x)
}

ggplot(multimappers, aes(x = "", y = count, fill = paste(quality, multiMapping))) + 
  geom_bar(stat = 'identity', position = position_fill())  +
  coord_polar(theta = "y")+
  facet_wrap(Method~factor(Sample, levels = c("Normal", "ccRCC_2", "ccRCC_3", "ccRCC_4", "ccRCC_5")), ncol=5)+
  theme_void() +
  scale_fill_manual(values = c("red", "darkcyan", "black", "grey")) + 
  geom_text(aes(label = ifelse(Percentage < 1, "", round(as.integer(Percentage),2))),
              position = position_fill(vjust = 0.5), color = "yellow", size = 10) +
  labs(fill = "Category") + ggtitle("Common cell barcode-UMI combinations") +
  theme(text = element_text(size = 25), legend.text = element_text(size = 15))
```


Why are the reads discarded in PacBio

```{r, fig.width=25, fig.height=10, echo=FALSE, message=FALSE, warning=FALSE}
ggplot() + geom_bar(data= multimappers, aes(x = "", y = Percentage, fill = factor(paste(quality, multiMapping), levels = c("Counted Multi mapping", "Counted Uniquely mapping", "NotCounted Multi mapping","NotCounted Uniquely mapping")), width = 10), stat = 'identity')  +
  coord_polar(theta = "y") +
  facet_wrap(Method~factor(Sample, levels = c("Normal", "ccRCC_2", "ccRCC_3", "ccRCC_4", "ccRCC_5")), ncol=5, scales = "free") + theme_classic() + 
    theme(axis.line = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          plot.title = element_text(hjust = 0.5, color = "#666666"), 
          strip.text = element_text(size = 13), 
          legend.text = element_text(size = 13)) +
  labs(fill = "Category") + 
  geom_bar(data = sub_data, aes(x = "", y = proportion, fill = quality2), stat = "identity", width = 3) + 
  facet_grid(Method~factor(Sample, levels = c("Normal", "ccRCC_2", "ccRCC_3", "ccRCC_4", "ccRCC_5"))) +
  coord_polar("y") + theme_classic()  + 
    theme(axis.line = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          plot.title = element_text(hjust = 0.5, color = "#666666"), 
          strip.text = element_text(size = 13), 
          legend.text = element_text(size = 13), legend.position = "bottom", guides(fill=guide_legend(ncol=2))) +
  scale_fill_manual(values = c("red", "darkcyan", "black","grey", "orange", "purple", "brown", "blue")) + 
  geom_text(data = multimappers, aes(x = "", y = Percentage, fill = factor(paste(quality, multiMapping), levels = c("Counted Multi mapping", "Counted Uniquely mapping", "NotCounted Multi mapping","NotCounted Uniquely mapping")), label = ifelse(Percentage < 1, "", round(as.integer(Percentage),0))),
              position = position_stack(vjust = 0.5), color = "yellow", size = 10) +
  theme(text = element_text(size = 30), legend.text = element_text(size = 26), strip.text = element_text(size = 27), legend.key.size = unit(1.5, "cm")) + labs(y = "", x = "")
```


```{r, fig.width=10, fig.height=15, echo=FALSE, message=FALSE, warning=FALSE}
ggplot() + geom_bar(data= multimappers, aes(x = "", y = Percentage, fill = factor(paste(quality, multiMapping), levels = c("Counted Multi mapping", "Counted Uniquely mapping", "NotCounted Multi mapping","NotCounted Uniquely mapping")), width = 10), stat = 'identity')  +
  coord_polar(theta = "y") +
  facet_wrap(factor(Sample, levels = c("Normal", "ccRCC_2", "ccRCC_3", "ccRCC_4", "ccRCC_5")) ~ Method, ncol=5, scales = "free") + theme_classic() + 
    theme(axis.line = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          plot.title = element_text(hjust = 0.5, color = "#666666"), 
          strip.text = element_text(size = 13), 
          legend.text = element_text(size = 13)) +
  labs(fill = "Category") + 
  geom_bar(data = sub_data, aes(x = "", y = proportion, fill = quality2), stat = "identity", width = 3) + 
  facet_grid(factor(Sample, levels = c("Normal", "ccRCC_2", "ccRCC_3", "ccRCC_4", "ccRCC_5")) ~ Method) +
  coord_polar("y") + theme_classic()  + 
    theme(axis.line = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          plot.title = element_text(hjust = 0.5, color = "#666666"), 
          strip.text = element_text(size = 13), 
          legend.text = element_text(size = 13), guides(fill=guide_legend(ncol=1))) +
  scale_fill_manual(values = c("red", "darkcyan", "black","grey", "orange", "purple", "brown", "blue")) + 
  geom_text(data = multimappers, aes(x = "", y = Percentage, fill = factor(paste(quality, multiMapping), levels = c("Counted Multi mapping", "Counted Uniquely mapping", "NotCounted Multi mapping","NotCounted Uniquely mapping")), label = ifelse(Percentage < 1, "", round(Percentage,0))),
              position = position_stack(vjust = 0.5), direction = "x", color = "yellow", size = 8) +
  theme(text = element_text(size = 15), legend.text = element_text(size = 15), strip.text = element_text(size = 20), legend.key.size = unit(1.5, "cm")) + labs(y = "", x = "")
```



```{r, fig.width=20, fig.height=15}
overlap = NULL
for (i in sample_data$Name){
  y = data.frame(illGaDF[[i]][illGaDF[[i]]$pbStatus == "pbDetected",]) %>% 
    group_by(NH, umiCount, overlapsPbAlign) %>% 
    dplyr::summarise(count = n_distinct(tagId)) %>% filter(!is.na(overlapsPbAlign))  %>%
    mutate(multiMapping = if_else(NH == 1, "Uniquely mapping", "Multi mapping"), quality = umiCount) %>% 
    group_by(multiMapping, quality, overlapsPbAlign) %>% 
    dplyr::summarise(count = sum(count)) %>% mutate(Percentage = (count * 100) / sum(count), Sample = i, category = paste(quality, multiMapping), overlapsPbAlign = if_else(overlapsPbAlign == "hasOverlapPb", "same location as PacBio", "different location than PacBio"))
  overlap = rbind(overlap, y)
}

ggplot(overlap, aes(x = "", y = count, fill = overlapsPbAlign)) + 
  geom_bar(stat = 'identity', position = position_fill())  + 
  facet_wrap(category~factor(Sample, levels = c("Normal", "ccRCC_2", "ccRCC_3", "ccRCC_4", "ccRCC_5")), ncol=5) +
  coord_polar(theta = "y") + 
  theme_void() + 
  labs(fill = "Category") + ggtitle("Common cell barcode-UMI combinations") +
  theme(text = element_text(size = 10), legend.text = element_text(size = 10)) +
  geom_text(aes(label = ifelse(Percentage < 0.09, "", round(Percentage,1))),
            position = position_fill(vjust = 0.5), color = "yellow", size = 7) + 
  scale_fill_manual(values = c("#00BFC4", "#F8766D"))
```



```{r, fig.width=10, fig.height=6}
ggplot(overlap, aes(category, Percentage, fill = overlapsPbAlign)) + 
  geom_bar(position = "fill", stat = "identity") + 
  facet_grid(~factor(Sample, levels = sample_data$Name)) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, color = c("red", "darkcyan", "black", "darkgrey"), size = 10), strip.text = element_text(size = 13), axis.text.y = element_text(size = 12), legend.position = "top", legend.text = element_text(size = 15)) +
  scale_fill_manual(values = c("#A6761D", "#440154FF")) + labs(x  = "", fill = "Mapping location", y = "proportion")
```

```{r, fig.width=10, fig.height=15}
ggplot(overlap, aes(category, Percentage, fill = overlapsPbAlign)) + 
  geom_bar(position = "fill", stat = "identity") + 
  facet_wrap(~factor(Sample, levels = sample_data$Name), ncol = 2) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, color = c("red", "darkcyan", "black", "darkgrey"), size = 15), strip.text = element_text(size = 13), axis.text.y = element_text(size = 12), legend.position = "top", legend.text = element_text(size = 15)) +
  scale_fill_manual(values = c("#A6761D", "#440154FF")) + labs(x  = "", fill = "Mapping location", y = "proportion")
```



```{r, fig.width=10, fig.height=5}
ggplot(overlap, aes(overlapsPbAlign, count, fill = category)) + 
    geom_bar(position = "fill", stat = "identity") + 
    facet_grid(~factor(Sample, levels = sample_data$Name)) +
    scale_fill_manual(values = c("red", "darkcyan", "black", "grey")) + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 11),strip.text = element_text(size = 13), axis.text.y = element_text(size = 12)) + labs(x = "Illumina alignment", y = "proportion")
```

Annotation of common tagids


```{r, fig.width=20, fig.height=7}
classification = NULL
for (i in sample_data$Name){
  x = pbGaDF[[i]]
  y = illGaDF[[i]]
  y = data.frame(y)[,c("tagId", "RE", "pbStatus")] %>% unique()
  x = data.frame(x)[,c("tagId", "category", "illStatus", "illUnmapped")] %>% unique()
  x = x %>% mutate(illStatus = if_else(illStatus == "illAbsent" & !is.na(illUnmapped), "illUnmapped", illStatus)) %>% dplyr::select("tagId", "category", "illStatus")
  xy = merge(x, y, by = "tagId", all = TRUE)  %>% mutate(category = if_else(is.na(category), "undetected", category), RE = case_when(is.na(RE) & illStatus != "illUnmapped" ~ "undetected", is.na(RE) & illStatus == "illUnmapped" ~ "Unmapped", .default = RE)) %>% dplyr::select(tagId, category, RE)
  xy = xy %>% group_by(category, RE) %>% dplyr::summarise(count = n()) %>% pivot_wider(names_from = RE, values_from = count) %>% replace(is.na(.), 0) %>% mutate(category = case_when(category == "full-splice_match" ~ "FSM", category == "incomplete-splice_match" ~ "ISM", category == "novel_in_catalog" ~ "NIC", category == "novel_not_in_catalog" ~ "NNC", .default = category)) %>% column_to_rownames("category") %>% rownames_to_column("PacBio") %>% melt() %>% mutate(Sample = i)
  classification = rbind(classification,xy)
}

classification = classification[classification$PacBio %in% c("FSM", "ISM", "NNC", "NIC", "genic", "unannotated", "intergenic", "undetected"),]
classification$PacBio = factor(classification$PacBio, levels = c("FSM", "ISM", "NNC", "NIC", "genic", "unannotated", "intergenic", "undetected"))
classification$Sample = factor(classification$Sample, levels = sample_data$Name)
classification = classification %>% mutate(variable = case_when(variable == "E" ~ "Exon", variable == "I" ~ "Intergenic", variable == "N" ~ "Intron" , variable == "undetected" ~ "undetected", variable == "Unmapped" ~ "Unmapped"))

ggplot(data = classification[!(classification$PacBio == "undetected" & classification$variable == "undetected"),] %>% mutate(value2 = round(value/100, 1)), 
       mapping = aes(x = variable, y = PacBio, fill = log(value+ 1))) + 
    geom_tile() +
    scale_fill_gradient2(high = "red", mid = "white", low = "blue", midpoint = 9, name = "Log(Abundance)") +
    geom_text(aes(label = value2), size = 2.7) +
    facet_grid(~Sample, switch = "x", scales = "free_x", space = "free_x") +
    theme_bw() + labs(x = "Illumina") + theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1), legend.position = "top")
```


```{r, fig.width=20, fig.height=7}
ggplot(data = classification[!(classification$PacBio == "undetected" & classification$variable == "undetected") & !(classification$PacBio == "undetected" & classification$variable == "Unmapped"),] %>% mutate(value2 = round(value/100, 1)), 
       mapping = aes(x = variable, y = PacBio, fill = log(value+ 1))) + 
    geom_tile() +
    scale_fill_viridis(name = "Log(Abundance)") +
    geom_text(aes(label = value2), size = 5) +
    facet_grid(~Sample, switch = "x", scales = "free_x", space = "free_x") +
    theme_bw() + labs(x = "Illumina") + theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1), legend.position = "top")
```


```{r, fig.width=7, fig.height=10}
ggplot(data = classification[!(classification$PacBio == "undetected" & classification$variable == "undetected") & !(classification$PacBio == "undetected" & classification$variable == "Unmapped"),] %>% mutate(value2 = round(value/100, 1)), 
       mapping = aes(x = variable, y = PacBio, fill = log(value+ 1))) + 
    geom_tile() +
    scale_fill_viridis(name = "Log(Abundance)") +
    geom_text(aes(label = value2), size = 2.7) +
    facet_wrap(~Sample, strip.position = "bottom", ncol = 2) +
    theme_bw() + labs(x = "Illumina") + theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1), legend.position = "top")
```

## Correlation between Illumina and Pacbio divided by width and GC content

```{r, fig.width=11, fig.height=10}
correlation = NULL
for (i in sample_data$Name){
  pbGaDF[[i]]$gcClasses = ezCut(pbGaDF[[i]]$gc, breaks = seq(0.3, 0.7, 0.1))
  pbGaDF[[i]]$widthClasses = ezCut(log2(pbGaDF[[i]]$qlength), breaks = 7:11)
  df = data.frame(pbGaDF[[i]])
  correlation = rbind(correlation, df)
}


for (i in sample_data$Name){
  df = data.frame(pbGaDF[[i]])
  illFrac <- tapply(as.integer(df$illStatus == "illDetected"), INDEX = list(width=df$widthClasses,
                                                                            gc=df$gcClasses), FUN = mean)
  corrplot::corrplot(illFrac, is.corr=FALSE, method = "pie", title = "Fraction of Pacbio reads common with Illumina", xlab="GC", ylab="log2 width", mar=c(0,2,2,0))
  mtext("GC content", side = 3, line=1)
  mtext("Log2 of read width", side = 2, line=2)
}
```

```{r, fig.width=11, fig.height=10}

illFrac <- tapply(as.integer(correlation$illStatus == "illDetected"), INDEX = list(width=correlation$widthClasses, gc=correlation$gcClasses), FUN = mean)
corrplot::corrplot(illFrac, is.corr=FALSE, method = "pie", title = "Fraction of Pacbio reads common with Illumina", xlab="GC", ylab="log2 width", mar=c(0,2,2,0), tl.col = "black")
mtext("GC content", side = 3, line=1)
mtext("Log2 of read width", side = 2, line=2)

```

## TSO and PolyA reads

```{r, fig.width=11, fig.height=4}
df = bind_rows(tso, .id = "Sample")
df$contaminants = case_when(df$ts == 1 & is.na(df$pa) ~ "TSO", df$pa == 1 & is.na(df$ts) ~ "polyA", df$ts == 1 & df$pa == 1 ~ "TSO+polyA", .default = "None")
df$InPB = if_else(df$InPB == "InPB", "Common", "Illumina specific")
df$Sample = factor(df$Sample, levels = sample_data$Name)
ggplot(df, aes(InPB, count, fill = contaminants)) + geom_bar(stat = "identity", position = "fill") + facet_grid(~Sample) + scale_fill_manual(values = myColors) + labs(x = "Cell category", y = "Proportion of reads") + theme(strip.text = element_text(size = 12), axis.text.y = element_text(size = 12), axis.text.x = element_text(size = 11, angle = 45, vjust = 1, hjust =1 ), legend.text = element_text(size = 12))
```